pypowerwall/tedapi/__init__.py

```diff
@@ -95,9 +98,11 @@ def uses_api_lock(func):
         func.api_lock = threading.Lock()
     @wraps(func)
     def wrapper(*args, **kwargs):
-        # Inject the function object itself into kwargs.
-        kwargs['self_function'] = func
-        return func(*args, **kwargs)
+        # Wrap the function with the lock
+        self = args[0]
+        with acquire_lock_with_backoff(func, self.timeout):
```

Comment:
If I understand correctly, this is a subtle change that wraps the ENTIRE function in the api lock, versus the specific parts of the function that are intended to be in a critical section. This reduces flexibility as there may be parts of each function that do not need to be locked.

Reply:
I'm leaning towards the trade-off that all the functions that were there right now are locking the whole thing, so it's probably better to make the decorator assume that and then if a function comes up later that only locks part that that would be the special case.

I think this is a bit more clear than having the extra `self_function` arg on every function, especially if you're ingesting it from the outside and using your IDE's function hinting.

Reply:
The `self_function` thing was indeed awkward. It was my way to workaround the fact that there's no built-in self-referential variable for a function, but you're absolutely right that it reduces the readability and makes it harder to use.

How about a compromise: Add a warning to the comments of the decorator (and the others) that note that this affects the behavior of the entire function and that it doesn't support locking part of the function.

Reply:
Fun story, I went to go compare some firmware stuff against the `main` branch and found that if I use `test_tedapi.py` and drop in a `gw.get_firmware_version()` throws an error because of the `self_function` argument.